[LAA] Strip findForkedPointer (NFC)

artagnon opened the pull request:
Remove a level of indirection due to findForkedPointer, in an effort to improve code.

fhahn commented on llvm/test/Transforms/LoopVectorize/pointer-select-runtime-checks.ll:
```diff
@@ -16,8 +16,10 @@ define void @test1_select_invariant(ptr %src.1, ptr %src.2, ptr %dst, i1 %c, i8
 ; CHECK-NEXT:    [[UGLYGEP1:%.*]] = getelementptr i8, ptr [[PTR_SEL]], i64 1
-; CHECK-NEXT:    [[BOUND0:%.*]] = icmp ult ptr [[DST]], [[UGLYGEP1]]
-; CHECK-NEXT:    [[BOUND1:%.*]] = icmp ult ptr [[PTR_SEL]], [[UGLYGEP]]
+; CHECK-NEXT:    [[PTR_SEL_FR:%.*]] = freeze ptr [[PTR_SEL]]
+; CHECK-NEXT:    [[SCEVGEP1_FR:%.*]] = freeze ptr [[UGLYGEP1]]
```
Are those freeze's needed? PTR_SEL is guaranteed to be dereferenced in the loop, so cannot be poison

fhahn commented on llvm/test/Transforms/LoopVectorize/pointer-select-runtime-checks.ll:
```diff
@@ -169,16 +171,12 @@ define void @test_loop_dependent_select2(ptr %src.1, ptr %src.2, ptr %dst, i8 %n
-; CHECK-NEXT:    [[BOUND03:%.*]] = icmp ult ptr [[DST]], [[UGLYGEP2_FR]]
-; CHECK-NEXT:    [[BOUND14:%.*]] = icmp ult ptr [[SRC_2_FR]], [[UGLYGEP]]
+; CHECK-NEXT:    [[BOUND03:%.*]] = icmp ult ptr [[DST]], [[UGLYGEP2]]
+; CHECK-NEXT:    [[BOUND14:%.*]] = icmp ult ptr [[SRC_2]], [[UGLYGEP]]
```
Freezes of SRC_1/SRC_2 should be needed, as either may be poison and not dereferenced in the loop, so w/o freeze we could introduce new branch on poison?

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -1088,29 +1092,49 @@ static void findForkedSCEVs(
-
-  // For now, we will only accept a forked pointer with two possible SCEVs
-  // that are either SCEVAddRecExprs or loop invariant.
-  if (Scevs.size() == 2 &&
```
Does the patch strip the requirement for at most 2 forked SCEVs?

artagnon replied:
No, not at all. If you see the original code after this `if`, it is:

```cpp
return {{replaceSymbolicStrideSCEV(PSE, StridesMap, Ptr), false}};
```

... which isn't guarded by size() == 1, and it inadvertently drops freeze.

huntergr-arm replied:
Looking through the tests, there's at least 3 which could generate more than 2 SCEV values (`@forked_ptrs_with_different_base3`, `@forked_ptrs_two_forks_gep`, and `@forked_ptrs_two_select`). However, the checks in `findForkedSCEVs` reject all 3 and just return a single CouldNotCompute SCEV.

I think this was just a final sanity check to restrict to 2 SCEVs rather than code we could properly test, in case someone extended `findForkedSCEVs` to cover new instructions without upholding the restriction.

There's no hard requirement to limit to 2 SCEVs of course, I was just focused on the cases I had at the time.

artagnon reviewed (commented):
Have to think about this a little bit, because I'm not sure what the correct freeze logic should be, and if findForkedSCEVs is setting freeze incorrectly. The previous code's freeze logic seems clearly incorrect though.

artagnon commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -1088,29 +1088,49 @@ static void findForkedSCEVs(
+  stable_sort(ForkedSCEVs, FreezeLess);
+  auto UniqPtrs = make_range(ForkedSCEVs.begin(), unique(ForkedSCEVs, PtrEq));
+
+  if (size(UniqPtrs) == 1) {
+    // FIXME: Is this correct?
+    UniqPtrs.begin()->setInt(false);
+    return UniqPtrs;
+  }
```
Okay, let's figure out the freeze issues in a follow-up. This patch is now an NFC, although it's quite hard to prove that.

artagnon replied:
Okay, I thought about it, and I think this is the correct logic.

artagnon commented:
Gentle ping.

artagnon commented:
Gentle ping. I think findForkedSCEVs can be made significantly more powerful, and I think that LAA will be able to permit a lot more cases for vectorization as a consequence. This patch is a prerequisite for that work.

igogo-x86 commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -1238,42 +1255,25 @@ bool AccessAnalysis::createCheckForAccess(
+      getRTCheckPtrs(PSE, TheLoop, Ptr, ForkedSCEVs, StridesMap, Assume);
+
+  /// Check whether all pointers can participate in a runtime bounds check: they
+  /// must either be loop-invariant, or an affine AddRec that does not wrap.
```
This comment doesn’t match the current behavior of the check

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -1207,29 +1207,46 @@ static void findForkedSCEVs(
+
+    // Call replaceSymbolicStrideSCEV only after PSE.getAsAddRec, because
+    // assumptions might have been added to PSE, resulting in simplifications.
+    const SCEV *S = replaceSymbolicStrideSCEV(PSE, StridesMap, Ptr);
```
Conceptually I am not sure if it make sense to call `replaceSymbolicStrideSCEV`; the pointer underlying IR pointer for the SCEV expression we are processing may be different than the original IR pointer, so using it here seems fragile for forked pointers.

fhahn reviewed (commented):
Taking a step back, it might be good to revisit the goal of the PR. Overall it looks to me like the new code isn't much simpler and adds more indirection.

How about just combining `findForkedPointer` and `findForkedSCEVs`?

artagnon committed:
[LAA] Strip findForkedPointer (NFC)

Remove a level of indirection due to findForkedPointer, in an effort to
improve code.

artagnon commented:
> Taking a step back, it might be good to revisit the goal of the PR. Overall it looks to me like the new code isn't much simpler and adds more indirection.
> 
> How about just combining `findForkedPointer` and `findForkedSCEVs`?

Thanks, the patch is now a clear NFC improvement.

artagnon committed:
[LAA] Address review

artagnon committed:
[LAA] Fix some nits

fhahn reviewed (approved):
LGTM, thanks
